### PR TITLE
[IMPAC-745] Fix Cash Projection Crash and Refactor Highcharts

### PR DESCRIPTION
**Pull request by iseessel**
@xaun
Implemented what we discussed. Its a pretty big pull request, so let me know if you have any questions. Let me know when you want me to squash ( would rather squash when ready to merge, as keeping track of my commits helps me).

I think this is an improvement over what was there originally. Only thing is its a bit confusing when @options should be populated in constructor vs in setter. All the dynamic options under the template, I put into setter methods, because they seemed specific to line charts.

### CODE REVIEW

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 1]
```diff
@@ -43,30 +28,33 @@ angular
...
-          selected: (if _.get(xAxisOptions, 'min') then null else 0)
 
   todayUTC = moment().startOf('day').add(moment().utcOffset(), 'minutes')
 
   class Chart
-    constructor: (@id, @data = {}, @options = {})->
-      @_template = templates[@options.chartType]
+    constructor: (@id, @series = {}, @settings = {})->
+      @_template = templates[@settings.chartType]
+      @populateChartOptions(@series, @settings)
```
Comment (xaun) [inline src/services/highcharts-factory/highcharts-factory.svc.coffee]: Not a big fan of how this method is public, seems more like a private method, and I think everything that happens in that method could just be within the constructor block.

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 2]
```diff
@@ -43,30 +28,33 @@ angular
...
-      if _.isEmpty(@hc)
-        @hc = Highcharts.stockChart(@id, chartConfig)
-      else
-        @hc.update(chartConfig)
+    populateChartOptions: (series, settings) ->
+      # Every chart will have these options. The rest of the options are created in the setter methods.
+      template = @_template
+      formatters = @formatters(settings.currency)
+      todayMarker = @todayMarker(settings)
+      series = @populateSeries(series)
```
Comment (xaun) [inline src/services/highcharts-factory/highcharts-factory.svc.coffee]: Prefer `series = { series: series }`, I dont think we need a `populateSeries` public instance method here

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 3]
```diff
@@ -43,30 +28,33 @@ angular
...
-      @data = data if _.isObject(data)
-      angular.extend(@options, options)
-      chartConfig = angular.merge({}, @template(), @formatters(), @todayMarker())
-      if _.isEmpty(@hc)
-        @hc = Highcharts.stockChart(@id, chartConfig)
-      else
-        @hc.update(chartConfig)
+    populateChartOptions: (series, settings) ->
+      # Every chart will have these options. The rest of the options are created in the setter methods.
+      template = @_template
```
Comment (xaun) [inline src/services/highcharts-factory/highcharts-factory.svc.coffee]: If you move the contents of this `populateChartOptions` method to just the constructor block, as @_template is not being accessed anywhere, we can remove it from the `this` scope, and just use 
```
template = templates[@settings.chartType]
angular.merge({}, templates ... ... . .) 
```

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 4]
```diff
@@ -86,11 +74,11 @@ angular
             name = _.startCase _.trim name.toLowerCase().replace(/\s*projected\s*/, ' ')
           "<strong>#{date}</strong><br>#{name}: #{amount}"
 
-    todayMarker: ->
-      return {} unless @options.showToday
+    todayMarker: (settings) ->
```
Comment (xaun) [inline src/services/highcharts-factory/highcharts-factory.svc.coffee]: `todayMarker: (showToday, markerColor = 'rgba(0, 85, 255, 0.2)')` ?

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 5]
```diff
@@ -130,22 +118,48 @@ angular
...
-        useHTML: useHTML
-        labelFormatter: formatterCallback
-      })
+    addCustomLegend: (labelFormatter, useHTML = true, showLegend = true) ->
+      legend =
+        legend:
+          labelFormatter: labelFormatter
+          useHTML: useHTML
+          enabled: showLegend
+      @options = angular.merge(@options, legend)
```
Comment (xaun) [inline src/services/highcharts-factory/highcharts-factory.svc.coffee]: The `angular.merge` method is destructive. The object in argument 1 will have the following object args merged into it. So @options will be mutated. If you want non destructive merging, pass an empty hash as the first argument like so:
```
# Note that @options will not contain attributes from someOtherHash
myNewHash = angular.merge({}, @options, someOtherHash)
```
So here you could simply write:
```
angular.merge(@options, legend)
```

File: src/components/widgets/accounts-class-comparison/accounts-class-comparison.directive.coffee [hunk 6]
```diff
@@ -35,7 +35,6 @@ module.controller('WidgetAccountsClassComparisonCtrl', ($scope, $q, $filter, Cha
   w.initContext = ->
     $scope.isDataFound = angular.isDefined(w.content) && !_.isEmpty(w.content.summary) && !_.isEmpty(w.content.companies)
     if $scope.isDataFound
-      $scope.timePeriodInfoParams.histParams = w.metadata && w.metadata.hist_parameters
```
Comment (xaun) [inline src/components/widgets/accounts-class-comparison/accounts-class-comparison.directive.coffee]: Why is this removed?

File: src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee [hunk 7]
```diff
@@ -173,6 +184,14 @@ module.controller('WidgetAccountsCashProjectionCtrl', ($scope, $q, $filter, $tim
       continue if s.userOptions.linkedTo != series.name
       if series.visible then s.hide() else s.show()
 
+  addCustomHighChartOptions = () ->
+    highChartsFactory = new HighchartsFactory($scope.chartId(), w.content.chart.series, _highChartOptions)
+    highChartsFactory.addCustomLegend(legendFormatter)
+    highChartsFactory.addSeriesEvent('click', onClickBar)
+    highChartsFactory.addSeriesEvent('legendItemClick', onClickLegend)
+    highChartsFactory.addXAxisOptions(_highChartOptions.withZooming)
```
Comment (xaun) [inline src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee]: I like the setters, means widgets can really configure the charts with a nice api. 

Thought It feels weird giving the factory the full `_highChartOptions` on init, then accessing one of the attributes for the `addXAxisOptions` method. I wouldn't shy away from using this syntax:

```coffeescript
highChartsFactory.addXAxisOptions({
    defaults: w.metadata.xAxis
    callback: onZoom
})
```

It means I you dont have to go scrolling up and down the file looking for the options hash, keeping things together. I think we should move the whole options to this if possible, see my comment at the `_highChartOptions` declarations.

File: src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee [hunk 8]
```diff
@@ -33,6 +33,17 @@ module.controller('WidgetAccountsCashProjectionCtrl', ($scope, $q, $filter, $tim
   $scope.chartId = ->
     "cashProjectionChart-#{w.id}"
 
+  # == Setup Highchart Options ============================================================================
+
+  _highChartOptions = Object.freeze
+    chartType: 'line'
+    chartOnClickCallbacks: []
+    currency: w.metadata.currency
```
Comment (xaun) [inline src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee]: As the variable `_highChartsOptions` is being defined in the controller function level of this directive, it will be evaluation as soon as the function is invoked internally by angular. This means, I am quite confident, w.metadata will be undefined at that exact point in time. 

This is essentially being decared as a constant, but as it has dynamic data inside it, it needs to either be `$onInit` (still fakey with asyc data), or better `$onChanges`, or even better (less code), re-evaluated when `w.format` is called.

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee [hunk 9]
```diff
@@ -67,14 +67,21 @@ module.component('chartThreshold', {
...
     ctrl.saveKpi = ->
       return if ctrl.loading
       ctrl.loading = true
       params = targets: {}, metadata: {}
       params.targets[ctrl.kpi.watchables[0]] = [{
         "#{ctrl.kpiTargetMode}": parseFloat(ctrl.draftTarget.value)
       }]
-      return unless ImpacKpisSvc.validateKpiTargets(params.targets)
+
+      if !ImpacKpisSvc.validateKpiTargets(params.targets)
```
Comment (xaun) [inline src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: prefer `unless` to `if !` with coffee

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee [hunk 10]
```diff
@@ -67,14 +67,21 @@ module.component('chartThreshold', {
       , 100)
       return
 
+    handleInvalidAlertAmount = ->
+      toastr.error("Please choose a number one or greater.", 'Error')
+      ctrl.cancelCreateKpi()
```
Comment (xaun) [inline src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: Is this new functionality? Do we want to `cancelCreateKpi` if I enter a wrong number?

File: src/services/highcharts-factory/highcharts-factory.svc.coffee (hunk 2 above)
Comment (iseessel) [reply src/services/highcharts-factory/highcharts-factory.svc.coffee]: Sure.

File: src/components/widgets/accounts-class-comparison/accounts-class-comparison.directive.coffee (hunk 6 above)
Comment (iseessel) [reply src/components/widgets/accounts-class-comparison/accounts-class-comparison.directive.coffee]: Oops most likely a rebasing problem.

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee (hunk 9 above)
Comment (iseessel) [reply src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: Sure.

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee (hunk 10 above)
Comment (iseessel) [reply src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: Yes we do, currently when we enter a wrong number, the update box will keep on spinning.

File: src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee (hunk 8 above)
Comment (iseessel) [reply src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee]: Good catch. Will put back into w.format.

File: src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee (hunk 7 above)
Comment (iseessel) [reply src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee]: Good point. What exactly do you mean by "we should move the whole options to this if possible."

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 11]
```diff
@@ -43,30 +28,33 @@ angular
...
             { type: 'all', text: 'All' }
           ]
-          selected: (if _.get(xAxisOptions, 'min') then null else 0)
 
   todayUTC = moment().startOf('day').add(moment().utcOffset(), 'minutes')
 
   class Chart
-    constructor: (@id, @data = {}, @options = {})->
-      @_template = templates[@options.chartType]
+    constructor: (@id, @series = {}, @settings = {})->
```
[the remark this answers is not in the document]
Comment (iseessel) [reply src/services/highcharts-factory/highcharts-factory.svc.coffee]: Yes, as can id.

File: src/services/highcharts-factory/highcharts-factory.svc.coffee (hunk 1 above)
Comment (iseessel) [reply src/services/highcharts-factory/highcharts-factory.svc.coffee]: Sure thing, can just move code over.

File: src/services/highcharts-factory/highcharts-factory.svc.coffee (hunk 4 above)
Comment (iseessel) [reply src/services/highcharts-factory/highcharts-factory.svc.coffee]: Yeah that works.

File: src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee (hunk 7 above)
Comment (xaun) [reply src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee]: @iseessel never-mind about that comment actually

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee (hunk 10 above)
Comment (xaun) [reply src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: @iseessel I had a quick look, the problem is the `ctrl.saveKpi` method sets `ctrl.loading = true`, but then returns without setting the `ctrl.loading` back to `false` unless `ImpacKpisSvc.validateKpiTargets(params.targets)`.

The quick solution to use a toastr to notify the user is fine I guess, though ngMessages txt would probs be better. But instead of calling `ctrl.cancelCreateKpi()` in your `handleInvalidAlertAmount` method, can we just try setting `ctrl.loading` to `false`? I think it would be a better ux.

File: src/services/highcharts-factory/highcharts-factory.svc.coffee (hunk 11 above)
[the remark this answers is not in the document]
Comment (iseessel) [reply src/services/highcharts-factory/highcharts-factory.svc.coffee]: EDIT: We can't remove settings from this, as it is called in #onChartNotify in `chart-threshold.component.coffee`

[the remark this answers is not in the document]
Comment (xaun) [reply src/services/highcharts-factory/highcharts-factory.svc.coffee]: Ah ok, do you think we should rename `@options` to `@hcOptions` or `@highchartsOptions` for clarity between options & settings?

[the remark this answers is not in the document]
Comment (iseessel) [reply src/services/highcharts-factory/highcharts-factory.svc.coffee]: Yeah probably a good idea.

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee (hunk 10 above)
Comment (iseessel) [reply src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: agreed

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 12]
```diff
@@ -4,37 +4,22 @@ angular
...
-        zoomingOptions = _.get(options, 'withZooming')
-        xAxisOptions = if zoomingOptions?
-          {
-            events:
-              setExtremes: zoomingOptions.callback
-            max: _.get(zoomingOptions.defaults, 'max')
-            min: _.get(zoomingOptions.defaults, 'min')
-          }
-
         chart:
```
Comment (xaun) [inline src/services/highcharts-factory/highcharts-factory.svc.coffee]: @iseessel there is inconsistent indentation here. The `line` object properties are indented 4 spaces, and the rest of the file is 2 spaces.

**Comment by xaun** [comment]
@iseessel if you could squash all commits like "refactoring", "debugging", "fix rebase problem",  "comments", "take out debugging". Commits like "Separated multiple update calls into one create call" can remain. Not sure why you need to keep "take out debuggers" commits during your  I development process? I suggest always using `git add -p` to stage your changes in hunks, giving you the chance to audit your work in pieces before committing, therefor not needing to remove debuggers later, and squash "remove debuggers" commits.

**Comment by iseessel** [comment]
@xaun Found a bug where an onclick callback did not work. Was quite a doozy to find the problem. 

It boiled down to a pointer program. You need the chartOnClickCallbacks to point to settings.onClickCallbacks, because in `chart-threshold-component.coffee' you push another callback into this array.

**Comment by iseessel** [comment]
@xaun So add manual transaction does not work in the cash projection widget, getting a 400 error from the server -- this is an existing error (tested it pre changes as well) and I suggest putting it in a new issue.

**Comment by xaun** [comment]
@iseessel I think there may be a better way that follows the new concept of setting options then rendering / re-rendering. 

If you look [here](https://github.com/iseessel/impac-angular/blob/d66bf5f814dea5c0a840da0682d8e8319ceeed9f/src/components/widgets-common/chart-threshold/chart-threshold.component.coffee#L44) you can see this is where the chart-threshold.component is calling `onChartNotify`. 

`ctrl.chartPromise` is a component binding given by the parent widget.

If we take a look [here](https://github.com/iseessel/impac-angular/blob/d66bf5f814dea5c0a840da0682d8e8319ceeed9f/src/components/widgets/accounts-cash-projection/accounts-cash-projection.directive.coffee#L214) at the cash-projection.directive, (`$scope.chartDeferred` is bound to chart-threshold.component as `chartPromise`), the notify is being invoked after render. I think if we simply invoke `$scope.chartDeferred.notify($scope.chart)` before calling `$scope.chart.render()`, it will work.
